velox/connectors/hive/storage_adapters/s3fs/RegisterS3FileSystem.cpp

```diff
@@ -26,73 +29,93 @@
 namespace facebook::velox::filesystems {
 
 #ifdef VELOX_ENABLE_S3
-folly::once_flag S3FSInstantiationFlag;
-
-// Only one instance of S3FileSystem is supported for now.
-// TODO: Support multiple S3FileSystem instances using a cache
-static std::shared_ptr<S3FileSystem> s3fs = nullptr;
+std::mutex s3fsmutex;
+/// Multiple S3 filesystems are supported.
+/// Key is the endpoint value specified in the config using hive.s3.endpoint.
+/// If the endpoint is empty, it will default to AWS S3.
+/// We cannot use folly::ConcurrentHashMap to store the filesystem instances.
+/// This is because we need to delete the filesystems during
+/// finalizeS3FileSystem and folly::ConcurrentHashMap does not guarantee the
+/// entries are cleared on invoking clear().
+static std::unordered_map<std::string, std::shared_ptr<FileSystem>> filesystems;
```

Comment:
can we use folly::Synchronized instead of relying on the explicit lock? If we have more than one structure we can always wrap them in a struct. Having these locks across different data structures can be error prone. 

Could also add this to a `getFilesystem()` function with an static local variable. Similar to what we do for vector functions .